plugin/pkg/admission/podpreset/admission.go

```diff
@@ -103,216 +104,222 @@ func (c *podPresetPlugin) Admit(a admission.Attributes) error {
+	applyPodPresetsOnPod(pod, matchingPPs)
 
-		// merge in policy for Env
-		if pip.Spec.Env != nil {
-			for i, ctr := range pod.Spec.Containers {
-				r, err := mergeEnv(pip, ctr.Env)
-				if err != nil {
-					// add event to pod
-					c.addEvent(pod, pip, err.Error())
+	glog.Infof("applied podpresets successfully on Pod: %s ", pod.GetGenerateName())
 
-					return nil
-				}
-				pod.Spec.Containers[i].Env = r
-			}
-		}
+	return nil
+}
 
-		// merge in policy for EnvFrom
-		if pip.Spec.EnvFrom != nil {
-			for i, ctr := range pod.Spec.Containers {
-				r, err := mergeEnvFrom(pip, ctr.EnvFrom)
-				if err != nil {
-					// add event to pod
-					c.addEvent(pod, pip, err.Error())
-
-					return nil
-				}
-				pod.Spec.Containers[i].EnvFrom = r
-			}
-		}
+func filterPodPresets(list []*settings.PodPreset, pod *api.Pod) ([]*settings.PodPreset, error) {
+	var matchingPPs []*settings.PodPreset
 
-		// merge in policy for VolumeMounts
-		if pip.Spec.VolumeMounts != nil {
-			for i, ctr := range pod.Spec.Containers {
-				r, err := mergeVolumeMounts(pip, ctr.VolumeMounts)
-				if err != nil {
-					// add event to pod
-					c.addEvent(pod, pip, err.Error())
-
-					return nil
-				}
-				pod.Spec.Containers[i].VolumeMounts = r
-			}
+	for _, pp := range list {
+		selector, err := metav1.LabelSelectorAsSelector(&pp.Spec.Selector)
+		if err != nil {
+			return nil, fmt.Errorf("label selector conversion failed: %v for selector: %v", pp.Spec.Selector, err)
 		}
 
-		// merge in policy for Volumes
-		if pip.Spec.Volumes != nil {
-			r, err := mergeVolumes(pip, pod.Spec.Volumes)
-			if err != nil {
-				// add event to pod
-				c.addEvent(pod, pip, err.Error())
-
-				return nil
-			}
-			pod.Spec.Volumes = r
+		// check if the pod labels match the selector
+		if !selector.Matches(labels.Set(pod.Labels)) {
+			continue
 		}
+		glog.V(4).Infof("PodPreset %s matches pod %s labels", pp.GetName(), pod.GetName())
+		matchingPPs = append(matchingPPs, pp)
+	}
+	return matchingPPs, nil
+}
 
-		glog.V(4).Infof("PodPreset %s merged with pod %s successfully", pip.GetName(), pod.GetName())
+func safeToApplyPodPresetsOnPod(pod *api.Pod, podPresets []*settings.PodPreset) error {
+	var errs []error
 
-		// add annotation
-		if pod.ObjectMeta.Annotations == nil {
-			pod.ObjectMeta.Annotations = map[string]string{}
+	if _, err := mergeVolumes(pod.Spec.Volumes, podPresets); err != nil {
+		errs = append(errs, err)
+	}
+	for _, ctr := range pod.Spec.Containers {
+		if err := safeToApplyPodPresetsOnContainer(&ctr, podPresets); err != nil {
+			errs = append(errs, err)
 		}
+	}
+	return utilerrors.NewAggregate(errs)
+}
 
-		pod.ObjectMeta.Annotations[fmt.Sprintf("%s/podpreset-%s", annotationPrefix, pip.GetName())] = pip.GetResourceVersion()
+func safeToApplyPodPresetsOnContainer(ctr *api.Container, podPresets []*settings.PodPreset) error {
+	var errs []error
+	if _, err := MergeEnv(ctr.Env, podPresets); err != nil {
```

Comment:
Seems like mergeEnv is probably unintentionally exported; we should make it private unless it's used elsewhere